src/modules/csm/csm.py

```diff
@@ -0,0 +1,306 @@
+        1. Collect all the attestation duties of the network validators for the frame.
+        2. Calculate the performance of each validator based on the attestations.
+        3. Calculate the share of each CSM node operator excluding underperforming validators.
+    """
+
+    COMPATIBLE_CONTRACT_VERSIONS = [1]
+    COMPATIBLE_CONSENSUS_VERSIONS = [1]
+
+    report_contract: CSFeeOracleContract
+
+    def __init__(self, w3: Web3):
+        self.report_contract = w3.csm.oracle
+        self.state = State.load()
+        super().__init__(w3)
+
+    def refresh_contracts(self):
+        self.report_contract = self.w3.csm.oracle  # type: ignore
+
+    def execute_module(self, last_finalized_blockstamp: BlockStamp) -> ModuleExecuteDelay:
+        collected = self.collect_data(last_finalized_blockstamp)
+        if not collected:
+            logger.info(
+                {"msg": "Data required for the report is not fully collected yet. Waiting for the next finalized epoch"}
+            )
+            return ModuleExecuteDelay.NEXT_FINALIZED_EPOCH
+
+        # pylint:disable=duplicate-code
+        report_blockstamp = self.get_blockstamp_for_report(last_finalized_blockstamp)
+        if not report_blockstamp:
+            return ModuleExecuteDelay.NEXT_FINALIZED_EPOCH
+
+        self.process_report(report_blockstamp)
+        return ModuleExecuteDelay.NEXT_SLOT
+
+    @lru_cache(maxsize=1)
+    @duration_meter()
+    def build_report(self, blockstamp: ReferenceBlockStamp) -> tuple:
+        # NOTE: We cannot use `r_epoch` from the `current_frame_range` call because the `blockstamp` is a
+        # `ReferenceBlockStamp`, hence it's a block the frame ends at. We use `ref_epoch` instead.
+        l_epoch, _ = self.current_frame_range(blockstamp)
+        r_epoch = blockstamp.ref_epoch
+
+        self.state.validate(l_epoch, r_epoch)
+        self.state.status()
+
+        distributed, shares = self.calculate_distribution(blockstamp)
+        if not distributed:
+            logger.info({"msg": "No shares distributed in the current frame"})
+
+        # Load the previous tree if any.
+        root = self.w3.csm.get_csm_tree_root(blockstamp)
+        cid = self.w3.csm.get_csm_tree_cid(blockstamp)
+
+        if cid:
+            logger.info({"msg": "Fetching tree by CID from IPFS", "cid": repr(cid)})
+            last_tree = Tree.decode(self.w3.ipfs.fetch(cid))
+
+            if last_tree.root != root:
+                raise ValueError("Unexpected tree root got from IPFS dump")
+
+            logger.info({"msg": "Restored tree from IPFS dump", "root": repr(root)})
+            # Update cumulative amount of shares for all operators.
+            for v in last_tree.tree.values:
+                no_id, amount = v["value"]
+                shares[no_id] += amount
+
+        tree = self.make_tree(shares)
+        if tree:
+            cid = self.w3.ipfs.publish(tree.encode())
+            root = tree.root
```

Comment:
for CSM

Reply:
As discussed internally, we would keep it as is, since the `publish` method is backed by multiple retries and switching to a fallback provider. At the same time if the underlying providers are not available for some reason, it can be the case for all the members because of some global outage. But the availability of a tree is crucial for a report.